internal/cephfs/fsjournal.go

```diff
@@ -70,15 +81,44 @@ func checkVolExists(ctx context.Context, volOptions *volumeOptions, secret map[s
 	imageUUID := imageData.ImageUUID
 	vid.FsSubvolName = imageData.ImageAttributes.ImageName
 
-	_, err = getVolumeRootPathCeph(ctx, volOptions, cr, volumeID(vid.FsSubvolName))
-	if err != nil {
-		var evnf ErrVolumeNotFound
-		if errors.As(err, &evnf) {
+	var evnf ErrVolumeNotFound
+
+	if sID != nil || pvID != nil {
+		clone, cloneInfoErr := getcloneInfo(ctx, volOptions, cr, volumeID(vid.FsSubvolName))
+		if cloneInfoErr != nil {
+			if errors.As(cloneInfoErr, &evnf) {
+				err = j.UndoReservation(ctx, volOptions.MetadataPool,
+					volOptions.MetadataPool, vid.FsSubvolName, volOptions.RequestName)
+				return nil, err
+			}
+			return nil, err
+		}
+		if clone.Status.State == cephFSCloneInprogress {
+			return nil, ErrCloneInProgress{err: fmt.Errorf("clone is in progress for %v", vid.FsSubvolName)}
+		}
+		if clone.Status.State == cephFSCloneFailed {
+			err = purgeVolume(ctx, volumeID(vid.FsSubvolName), cr, volOptions, true)
+			if err != nil {
+				klog.Errorf(util.Log(ctx, "failed to delete volume %s: %v"), vid.FsSubvolName, err)
+				return nil, err
+			}
```

Comment:
Same as above, we delete the clone (as it failed), we should also call `cleanupCloneFromSubvolumeSnapshot` to ensure the intermediate snapshot is unprotected and deleted if this was a clone from a volume.